apps/remote_control/lib/lexical/remote_control/code_mod/rename/file.ex

```diff
@@ -0,0 +1,172 @@
+defmodule Lexical.RemoteControl.CodeMod.Rename.File do
+  alias Lexical.Ast
+  alias Lexical.Document
+  alias Lexical.ProcessCache
+  alias Lexical.Project
+  alias Lexical.RemoteControl
+  alias Lexical.RemoteControl.Search.Indexer
+  alias Lexical.RemoteControl.Search.Indexer.Entry
+
+  @spec maybe_rename(Entry.t(), String.t()) :: Document.Changes.rename_file()
+  def maybe_rename(entry, new_suffix) do
+    if root_module?(entry) do
+      rename_file(entry, new_suffix)
+    end
+  end
+
+  defp root_module?(entry) do
+    uri = Document.Path.ensure_uri(entry.path)
+
+    entries =
+      ProcessCache.trans("#{uri}-entries", 50, fn ->
+        with {:ok, document} <- Document.Store.open_temporary(uri),
+             {:ok, entries} <-
+               Indexer.Source.index_document(document, [Indexer.Extractors.Module]) do
+          entries
+        end
+      end)
+
+    case Enum.filter(entries, &(&1.block_id == :root)) do
+      [root_module] ->
+        root_module.subject == entry.subject and root_module.block_range == entry.block_range
+
+      _ ->
+        false
+    end
+  end
+
+  defp rename_file(entry, new_suffix) do
+    root_path = root_path()
+    relative_path = relative_path(entry.path, root_path)
+
+    with {:ok, prefix, old_module_paths} <- fetch_conventional_prefix(relative_path),
+         {:ok, new_name} <- fetch_new_name(entry, new_suffix) do
+      extname = Path.extname(entry.path)
+
+      suffix =
+        new_name
+        |> Macro.underscore()
+        |> maybe_insert_special_phoenix_folder(old_module_paths, new_name)
+
+      new_path = Path.join([root_path, prefix, "#{suffix}#{extname}"])
+
+      old_uri = Document.Path.ensure_uri(entry.path)
+      new_uri = Document.Path.ensure_uri(new_path)
+      Document.Changes.RenameFile.new(old_uri, new_uri)
+    else
+      _ -> nil
+    end
+  end
+
+  defp relative_path(path, root_path) do
+    Path.relative_to(path, root_path)
+  end
+
+  defp root_path do
+    Project.root_path(RemoteControl.get_project())
+  end
+
+  defp fetch_new_name(entry, new_suffix) do
+    uri = Document.Path.ensure_uri(entry.path)
+    text_edits = [Document.Edit.new(new_suffix, entry.range)]
+
+    with {:ok, document} <- Document.Store.open_temporary(uri),
+         {:ok, edited_document} =
+           Document.apply_content_changes(document, document.version + 1, text_edits),
+         {:ok, %{context: {:alias, alias}}} <-
+           Ast.surround_context(edited_document, entry.range.start) do
+      {:ok, to_string(alias)}
+    else
```

Comment:
This feels incredibly wasteful to me, you're opening a new document, applying content edits to it, then parsing it just to get the new name of a file. Couldn't you do something similar by pulling using the range of the document and modifying that?

Reply:
Perhaps we should pass the `document` params in  `maybe_rename` function, as this would simplify many things.